geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: http-route-cors
+spec:
+  hostnames:
+  - http.route.cors.com
+  parentRefs:
+  - group: gateway.networking.k8s.io
+    kind: Gateway
+    name: http-gateway
+  rules:
+  - backendRefs:
+    - kind: Service
+      name: http-route-cors
+      port: 80
+    matches:
+    - path:
+        type: PathPrefix
+        value: /resource/foo
+    filters:
+    - cors:
+      - allowOrigins:
+        - https://foo.example
+        - http://foo.example
+        allowMethods: 
+        - GET
+        - PUT
+        - POST
+        - DELETE
+        - PATCH
+        - OPTIONS
+        allowHeaders: 
+        - DNT
+        - X-CustomHeader
+        - Keep-Alive
+        - User-Agent
+        - X-Requested-With
+        - If-Modified-Since
+        - Cache-Control
+        - Content-Type
+        - Authorization
+        exposeHeaders: 
+        - Content-Security-Policy
+        maxAge: 1728000
+      type: CORS
+```
+
+A client sends a cross-origin "preflight" request for asking a server whether it would allow a PUT request before the actual cross-origin request is sent.
+```
+OPTIONS /resource/foo HTTP/1.1
+Host: http.route.cors.com
+Origin: https://foo.example
+Access-Control-Request-Method: PUT
+```
+
+The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+Based on the above HTTPRoute, a successful "preflight" response will be generated by the gateway. Moreover, the gateway will send the "preflight" response to the client directly.
+
+```
+HTTP/1.1 204 No Content
+Access-Control-Allow-Origin: https://foo.example
+Access-Control-Allow-Credentials: true
+Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS
+Access-Control-Allow-Headers: DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization
+Access-Control-Max-Age: 1728000
+Content-Type: text/plain charset=UTF-8
+Content-Length: 0
+```
+
+Then, the client will be able to send actual cross-origin request.
+```
+PUT /resource/foo HTTP/1.1
+Host: http.route.cors.com
+Keep-Alive: timeout=5, max=1000
+Origin: https://foo.example
+Authorization: Basic YWxhZGRpbjpvcGVuc2VzYW1l
+```
+
+At last, the cross-origin response headers will be added by the gateway to the response based on the HTTPRoute.
+```
+Access-Control-Allow-Origin: https://foo.example
+Access-Control-Allow-Credentials: true
+Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS
+Access-Control-Allow-Headers: DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization
+Access-Control-Expose-Headers: Content-Security-Policy
+```
+
+## Prior Art
```

Comment:
This section really should be before the API section and after the Introduction section. I think we may need a template update.